File: frontend/src/types/training.ts
import type { TrainingStatus } from "./components";

export type TrainingJobStatus =
  | "QUEUED"
  | "VALIDATING"
  | "PREPARING_DATA"
  | "RUNNING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELED";

export type TrainingJobPhase =
  | "REQUESTED"
  | "VALIDATION"
  | "DATA_PREPARATION"
  | "TRAINING"
  | "EVALUATION"
  | "PUBLISHING"
  | "FINISHED";

export interface TrainingJobRecipeSnapshot {
  recipe_id: string;
  name: string;
  model_key: string;
  dataset_profile: string;
  league_ids: string[];
  days_back: number;
  feature_profile: string;
  hyperparameter_profile: string;
  executor_target: string;
  publish_strategy: string;
  requested_by: string | null;
  requested_at: string | null;
  description: string | null;
}

export interface TrainingJobCreateRequest {
  recipe_id: string;
  name: string;
  model_key: string;
  dataset_profile: string;
  league_ids: string[];
  days_back: number;
  feature_profile?: string;
  hyperparameter_profile?: string;
  executor_target?: string;
  publish_strategy?: string;
  description?: string | null;
}

export interface TrainingJobSummary {
  job_id: string;
  status: TrainingJobStatus;
  status_message: string;
  progress_percent: number;
  model_key: string | null;
  executor_target: string | null;
  created_at: string | null;
  updated_at: string | null;
  phase?: TrainingJobPhase;
  executor_type?: string | null;
  executor_run_id?: string | null;
  recipe_snapshot?: TrainingJobRecipeSnapshot;
  result_summary?: Record<string, unknown>;
  artifact_ids?: string[];
  audit_trail?: Record<string, unknown>[];
  started_at?: string | null;
  finished_at?: string | null;
  cancel_requested_at?: string | null;
  error_code?: string | null;
  error_message?: string | null;
}

export interface TrainingJobEvent {
  event_id: string;
  job_id: string;
  event_type: string;
  message: string;
  phase: TrainingJobPhase | null;
  progress_percent: number | null;
  payload: Record<string, unknown>;
  created_at: string | null;
}

export interface TrainingOption {
  key: string;
  label: string;
  description?: string;
}

export interface TrainingModelOption extends TrainingOption {
  supported_feature_profiles: string[];
  supported_dataset_profiles: string[];
  supported_executor_targets: string[];
  supported_league_ids: string[];
  supported_days_back: number[];
  default_executor_target: string | null;
}

export interface TrainingExecutorOption extends TrainingOption {
  available: boolean;
  supports_cancel: boolean;
  supports_logs: boolean;
}

export interface TrainingCapabilityReason {
  code: string;
  message: string;
}

export interface TrainingCapabilities {
  available: boolean;
  models: TrainingModelOption[];
  executors: TrainingExecutorOption[];
  dataset_profiles: TrainingOption[];
  feature_profiles: TrainingOption[];
  league_options: TrainingOption[];
  days_back_options: number[];
  reasons: TrainingCapabilityReason[];
}

export interface TrainingLatestResult {
  available: boolean;
  data: TrainingStatus | null;
  last_update: string | null;
}

File: backend/src/api/schemas/training.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field
from src.domain.training.models import (
    ArtifactStatus,
    PublishStrategy,
    TrainingJobPhase,
    TrainingJobStatus,
)


class TrainingOptionModel(BaseModel):
    key: str
    label: str
    description: str | None = None


class TrainingModelOptionModel(TrainingOptionModel):
    supported_feature_profiles: list[str] = Field(default_factory=list)
    supported_dataset_profiles: list[str] = Field(default_factory=list)
    supported_executor_targets: list[str] = Field(default_factory=list)
    supported_league_ids: list[str] = Field(default_factory=list)
    supported_days_back: list[int] = Field(default_factory=list)
    default_executor_target: str | None = None


class TrainingExecutorOptionModel(TrainingOptionModel):
    available: bool = True
    supports_cancel: bool = False
    supports_logs: bool = False


class TrainingUnavailableReasonModel(BaseModel):
    code: str
    message: str


class TrainingCapabilitiesPayload(BaseModel):
    available: bool
    models: list[TrainingModelOptionModel] = Field(default_factory=list)
    executors: list[TrainingExecutorOptionModel] = Field(default_factory=list)
    dataset_profiles: list[TrainingOptionModel] = Field(default_factory=list)
    feature_profiles: list[TrainingOptionModel] = Field(default_factory=list)
    league_options: list[TrainingOptionModel] = Field(default_factory=list)
    days_back_options: list[int] = Field(default_factory=list)
    reasons: list[TrainingUnavailableReasonModel] = Field(default_factory=list)


class TrainingModelsPayload(BaseModel):
    models: list[TrainingModelOptionModel] = Field(default_factory=list)


class TrainingLatestResultPayload(BaseModel):
    available: bool
    data: dict[str, Any] | None = None
    last_update: str | None = None


class TrainingJobCreatePayload(BaseModel):
    recipe_id: str
    name: str
    model_key: str
    dataset_profile: str
    league_ids: list[str] = Field(default_factory=list)
    days_back: int = 0
    feature_profile: str = "default"
    hyperparameter_profile: str = "default"
    executor_target: str = "default"
    publish_strategy: PublishStrategy = PublishStrategy.MANUAL
    description: str | None = None


class TrainingJobPayload(BaseModel):
    job_id: str
    recipe_snapshot: dict[str, Any]
    status: TrainingJobStatus
    status_message: str
    progress_percent: int = 0
    phase: TrainingJobPhase
    executor_type: str | None = None
    executor_run_id: str | None = None
    queued_at: str | None = None
    started_at: str | None = None
    finished_at: str | None = None
    cancel_requested_at: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    result_summary: dict[str, Any] = Field(default_factory=dict)
    artifact_ids: list[str] = Field(default_factory=list)
    audit_trail: list[dict[str, Any]] = Field(default_factory=list)


class TrainingJobListPayload(BaseModel):
    jobs: list[TrainingJobPayload] = Field(default_factory=list)


class TrainingJobEventPayload(BaseModel):
    event_id: str
    job_id: str
    event_type: str
    message: str
    phase: TrainingJobPhase | None = None
    progress_percent: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None


class TrainingJobEventsPayload(BaseModel):
    job_id: str
    events: list[TrainingJobEventPayload] = Field(default_factory=list)


class ModelArtifactPayload(BaseModel):
    artifact_id: str
    job_id: str
    model_key: str
    status: ArtifactStatus
    metrics: dict[str, Any] = Field(default_factory=dict)
    feature_contract_version: str
    training_data_fingerprint: str | None = None
    context_summary: dict[str, Any] = Field(default_factory=dict)
    storage_location: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None
    published_at: str | None = None


class ActiveModelPointerPayload(BaseModel):
    scope_key: str
    artifact_id: str
    model_key: str
    promoted_by: str | None = None
    promoted_at: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class PromotionPayload(BaseModel):
    artifact_id: str
    scope_key: str
    promoted_by: str | None = None
